File: backend-scala/prism-core/src/main/scala/prism/core/Protocol.scala
package prism.core

import prism.core.Types.HydratedReference
import upickle.default._

/** Prism protocol message definitions.
  *
  * Defines the message types that flow between client and server over the transport layer.
  */
object Protocol {

  // ========== Client -> Server Messages ==========

  /** Subscribe to object updates. */
  final case class SubscribeMessage(
      objectId: String,
      filterType: Option[String] = None,
      filterParams: Option[Map[String, ujson.Value]] = None,
      temporary: Boolean = false
  )

  object SubscribeMessage {
    implicit val rw: ReadWriter[SubscribeMessage] = macroRW
  }

  /** Unsubscribe from object updates. */
  final case class UnsubscribeMessage(
      objectId: String
  )

  object UnsubscribeMessage {
    implicit val rw: ReadWriter[UnsubscribeMessage] = macroRW
  }

  /** Client's current state for an object during sync. */
  final case class SyncStateItem(
      id: String,
      version: Int,
      filterType: String
  )

  object SyncStateItem {
    implicit val rw: ReadWriter[SyncStateItem] = macroRW
  }

  /** Synchronize state after reconnection. */
  final case class SyncMessage(
      states: List[SyncStateItem]
  )

  object SyncMessage {
    implicit val rw: ReadWriter[SyncMessage] = macroRW
  }

  /** Business logic request with automatic hydration. */
  final case class RequestMessage(
      requestId: String,
      requestType: String,
      payload: ujson.Value,
      options: Option[ujson.Value] = None
  )

  object RequestMessage {
    implicit val rw: ReadWriter[RequestMessage] = macroRW
  }

  /** Update filter for existing subscription. */
  final case class UpdateFilterMessage(
      objectId: String,
      filterType: String,
      filterParams: Option[Map[String, ujson.Value]] = None
  )

  object UpdateFilterMessage {
    implicit val rw: ReadWriter[UpdateFilterMessage] = macroRW
  }

  /** Helper to merge ujson objects */
  private def mergeObj(base: ujson.Obj, other: ujson.Obj): ujson.Obj = {
    val result = ujson.Obj()
    base.value.foreach { case (k, v) => result(k) = v }
    other.value.foreach { case (k, v) => result(k) = v }
    result
  }

  /** Union type for all client messages. */
  sealed trait ClientMessage

  object ClientMessage {
    final case class Subscribe(msg: SubscribeMessage) extends ClientMessage
    final case class Unsubscribe(msg: UnsubscribeMessage) extends ClientMessage
    final case class Sync(msg: SyncMessage) extends ClientMessage
    final case class Request(msg: RequestMessage) extends ClientMessage
    final case class UpdateFilter(msg: UpdateFilterMessage) extends ClientMessage

    // Custom JSON reader/writer with discriminator
    implicit val rw: ReadWriter[ClientMessage] = readwriter[ujson.Value].bimap[ClientMessage](
      {
        case Subscribe(msg) =>
          mergeObj(ujson.Obj("type" -> "subscribe"), writeJs(msg).obj)
        case Unsubscribe(msg) =>
          mergeObj(ujson.Obj("type" -> "unsubscribe"), writeJs(msg).obj)
        case Sync(msg) =>
          mergeObj(ujson.Obj("type" -> "sync"), writeJs(msg).obj)
        case Request(msg) =>
          mergeObj(ujson.Obj("type" -> "request"), writeJs(msg).obj)
        case UpdateFilter(msg) =>
          mergeObj(ujson.Obj("type" -> "updateFilter"), writeJs(msg).obj)
      },
      json => {
        val msgType = json("type").str
        msgType match {
          case "subscribe"    => Subscribe(read[SubscribeMessage](json))
          case "unsubscribe"  => Unsubscribe(read[UnsubscribeMessage](json))
          case "sync"         => Sync(read[SyncMessage](json))
          case "request"      => Request(read[RequestMessage](json))
          case "updateFilter" => UpdateFilter(read[UpdateFilterMessage](json))
          case _              => throw new IllegalArgumentException(s"Unknown client message type: $msgType")
        }
      }
    )
  }

  // ========== Server -> Client Messages ==========

  /** Full object transmission. */
  final case class FullObjectMessage(
      id: String,
      version: Int,
      data: ujson.Value,
      filtered: Boolean = false,
      filterType: Option[String] = None
  )

  object FullObjectMessage {
    implicit val rw: ReadWriter[FullObjectMessage] = macroRW
  }

  /** Delta update for an object. */
  final case class DeltaMessage(
      id: String,
      fromVersion: Int,
      toVersion: Int,
      patches: List[ujson.Value],
      filterType: Option[String] = None
  )

  object DeltaMessage {
    implicit val rw: ReadWriter[DeltaMessage] = macroRW
  }

  /** Response to a request with smart hydration. */
  final case class ResponseMessage(
      requestId: String,
      success: Boolean,
      data: Option[ujson.Value] = None,
      hydrated: List[HydratedReference] = List.empty,
      error: Option[String] = None
  )

  object ResponseMessage {
    implicit val rw: ReadWriter[ResponseMessage] = macroRW
  }

  /** Error notification. */
  final case class ErrorMessage(
      code: String,
      message: String,
      objectId: Option[String] = None,
      requestId: Option[String] = None
  )

  object ErrorMessage {
    implicit val rw: ReadWriter[ErrorMessage] = macroRW
  }

  /** Union type for all server messages. */
  sealed trait ServerMessage

  object ServerMessage {
    final case class FullObject(msg: FullObjectMessage) extends ServerMessage
    final case class Delta(msg: DeltaMessage) extends ServerMessage
    final case class Response(msg: ResponseMessage) extends ServerMessage
    final case class Error(msg: ErrorMessage) extends ServerMessage

    // Custom JSON reader/writer with discriminator
    implicit val rw: ReadWriter[ServerMessage] = readwriter[ujson.Value].bimap[ServerMessage](
      {
        case FullObject(msg) =>
          mergeObj(ujson.Obj("type" -> "fullObject"), writeJs(msg).obj)
        case Delta(msg) =>
          mergeObj(ujson.Obj("type" -> "delta"), writeJs(msg).obj)
        case Response(msg) =>
          mergeObj(ujson.Obj("type" -> "response"), writeJs(msg).obj)
        case Error(msg) =>
          mergeObj(ujson.Obj("type" -> "error"), writeJs(msg).obj)
      },
      json => {
        val msgType = json("type").str
        msgType match {
          case "fullObject" => FullObject(read[FullObjectMessage](json))
          case "delta"      => Delta(read[DeltaMessage](json))
          case "response"   => Response(read[ResponseMessage](json))
          case "error"      => Error(read[ErrorMessage](json))
          case _            => throw new IllegalArgumentException(s"Unknown server message type: $msgType")
        }
      }
    )
  }

  // ========== Error Codes ==========

  object ErrorCodes {
    val OBJECT_NOT_FOUND = "OBJECT_NOT_FOUND"
    val INVALID_REQUEST = "INVALID_REQUEST"
    val UNAUTHORIZED = "UNAUTHORIZED"
    val INTERNAL_ERROR = "INTERNAL_ERROR"
    val INVALID_FILTER = "INVALID_FILTER"
    val INVALID_VERSION = "INVALID_VERSION"
  }
}

File: backend/prism/core/protocol.py
"""Prism protocol message definitions.

Defines the message types that flow between client and server over the transport layer.
"""

from typing import Annotated, Any, Literal

from pydantic import BaseModel, Discriminator, Field

from prism.core.types import HydratedReference

# Client -> Server Messages


class SubscribeMessage(BaseModel):
    """Subscribe to object updates."""

    type: Literal["subscribe"] = "subscribe"
    object_id: str
    filter_type: str | None = None
    filter_params: dict[str, Any] | None = None
    temporary: bool = Field(False, description="One-time fetch vs ongoing subscription")


class UnsubscribeMessage(BaseModel):
    """Unsubscribe from object updates."""

    type: Literal["unsubscribe"] = "unsubscribe"
    object_id: str


class SyncStateItem(BaseModel):
    """Client's current state for an object."""

    id: str
    version: int
    filter_type: str


class SyncMessage(BaseModel):
    """Synchronize state after reconnection."""

    type: Literal["sync"] = "sync"
    states: list[SyncStateItem]


class RequestMessage(BaseModel):
    """Business logic request with automatic hydration."""

    type: Literal["request"] = "request"
    request_id: str
    request_type: str
    payload: dict[str, Any]
    options: dict[str, Any] | None = None


class UpdateFilterMessage(BaseModel):
    """Update filter for existing subscription."""

    type: Literal["updateFilter"] = "updateFilter"
    object_id: str
    filter_type: str
    filter_params: dict[str, Any] | None = None


def get_client_message_type(v: Any) -> str:
    """Discriminator function for client messages."""
    if isinstance(v, dict):
        return str(v.get("type", "unknown"))
    return str(getattr(v, "type", "unknown"))


ClientMessage = Annotated[
    SubscribeMessage
    | UnsubscribeMessage
    | SyncMessage
    | RequestMessage
    | UpdateFilterMessage,
    Discriminator(get_client_message_type),
]


# Server -> Client Messages


class FullObjectMessage(BaseModel):
    """Full object transmission."""

    type: Literal["fullObject"] = "fullObject"
    id: str
    version: int
    data: dict[str, Any]
    filtered: bool = False
    filter_type: str | None = None


class DeltaMessage(BaseModel):
    """Delta update for an object."""

    type: Literal["delta"] = "delta"
    id: str
    from_version: int
    to_version: int
    patches: list[dict[str, Any]]
    filter_type: str | None = None


class ResponseMessage(BaseModel):
    """Response to a request with smart hydration."""

    type: Literal["response"] = "response"
    request_id: str
    success: bool
    data: dict[str, Any] | None = None
    hydrated: list[HydratedReference] = Field(default_factory=list)
    error: str | None = None


class ErrorMessage(BaseModel):
    """Error notification."""

    type: Literal["error"] = "error"
    code: str
    message: str
    object_id: str | None = None
    request_id: str | None = None


def get_server_message_type(v: Any) -> str:
    """Discriminator function for server messages."""
    if isinstance(v, dict):
        return str(v.get("type", "unknown"))
    return str(getattr(v, "type", "unknown"))


ServerMessage = Annotated[
    FullObjectMessage | DeltaMessage | ResponseMessage | ErrorMessage,
    Discriminator(get_server_message_type),
]
